pkg/common/workbuilder/workbuilder.go

```diff
@@ -0,0 +1,318 @@
+			totalSize = totalSize + manifests[i].Size()
+			continue
+		}
+
+		workIndex = workIndex + 1
+		work := w.generateManifestWork(workIndex)
+		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, manifests[i])
+		totalSize = manifests[i].Size()
+		manifestWorks = append(manifestWorks, work)
+	}
+
+	return manifestWorks, nil
+}
+
+func (w *WorkBuilder) buildManifestWorks(objects []runtime.Object,
+	existingManifestWorks []workapiv1.ManifestWork) (appliedWorks, deletedWorks []workapiv1.ManifestWork, err error) {
+	if len(existingManifestWorks) == 0 {
+		requiredWorks, err := w.NewBuild(objects)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to new build works. err: %v", err)
+		}
+		return requiredWorks, nil, nil
+	}
+
+	requireMapper, err := generateRequiredManifestMapper(objects)
+	if err != nil {
+		return nil, nil, err
+	}
+	return w.compareManifests(existingManifestWorks, requireMapper)
+}
+
+func generateManifestKey(manifest workapiv1.Manifest) (manifestKey, error) {
+	var object runtime.Object
+	var err error
+	if manifest.Object != nil {
+		object = manifest.Object
+	} else {
+		object, err = runtime.Decode(unstructured.UnstructuredJSONScheme, manifest.Raw)
+		if err != nil {
+			return manifestKey{}, err
+		}
+	}
+
+	key := manifestKey{
+		gvk: object.GetObjectKind().GroupVersionKind(),
+	}
+
+	accessor, err := meta.Accessor(object)
+	if err != nil {
+		return key, err
+	}
+	key.namespace = accessor.GetNamespace()
+	key.name = accessor.GetName()
+	return key, nil
+}
+
+func buildManifest(object runtime.Object) (workapiv1.Manifest, error) {
+	rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+	if err != nil {
+		return workapiv1.Manifest{}, fmt.Errorf("failed to encode object %v, err: %v", object, err)
+	}
+	manifest := workapiv1.Manifest{RawExtension: runtime.RawExtension{Raw: rawObject}}
+	return manifest, nil
+}
+
+func generateRequiredManifestMapper(objects []runtime.Object) (requiredManifestMapper, error) {
+	var mapper = requiredManifestMapper{}
+	for _, object := range objects {
+		rawObject, err := runtime.Encode(unstructured.UnstructuredJSONScheme, object)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode object %v, err: %v", object, err)
+		}
+		key := manifestKey{
+			gvk: object.GetObjectKind().GroupVersionKind(),
+		}
+		accessor, err := meta.Accessor(object)
+		if err != nil {
+			return nil, err
+		}
+		key.namespace = accessor.GetNamespace()
+		key.name = accessor.GetName()
+		mapper[key] = workapiv1.Manifest{RawExtension: runtime.RawExtension{Raw: rawObject}}
+	}
+	return mapper, nil
+}
+
+func (w *WorkBuilder) compareManifests(existingWorks []workapiv1.ManifestWork,
+	requiredMapper requiredManifestMapper) (appliedWorks, deletedWorks []workapiv1.ManifestWork, err error) {
+	var updatedWorks []workapiv1.ManifestWork
+	for workIndex := 0; workIndex < len(existingWorks); workIndex++ {
```

Comment:
add a comment here, this step is to update existing manifestwork, delete the non-existing manifests, update existing manifets.

Reply:
done